util/design/gen-flash-img.py

```diff
@@ -2,79 +2,129 @@
 # Copyright lowRISC contributors.
 # Licensed under the Apache License, Version 2.0, see LICENSE for details.
 # SPDX-License-Identifier: Apache-2.0
-r"""Takes a compiled vmem image and processes it for flash.
-    Long term this should include both layers of ECC and scrambling,
-    The first version has only the truncated plaintext ECC.
+r"""Takes a compiled VMEM image and processes it for loading into flash.
+
+    Specifically, this takes a raw flash image and adds both layers of ECC
+    (integrity and reliablity), and optionally, scrambles the data using the
+    same XEX scrambling scheme used in the flash controller. This enables
+    backdoor loading the flash on simulation platforms (e.g., DV and Verilator).
 """
 
 import argparse
-import math
+import functools
+import logging
 import re
+import sys
 from pathlib import Path
-from typing import Dict, Any
-
-import secded_gen
-
-
-def _add_intg_ecc(config: Dict[str, Any], in_val: int) -> str:
-    result, m = secded_gen.ecc_encode(config, "hamming", 64, in_val)
+from typing import List
 
-    m_nibbles = math.ceil(m / 4)
-    result = format(result, '0' + str(16 + m_nibbles) + 'x')
+from pyfinite import ffield
 
-    # due to lack of storage space, the first nibble of the ECC is truncated
-    return result[1:]
+import prince
+import secded_gen
 
+FLASH_WORD_SIZE = 64  # bits
+FLASH_ADDR_SIZE = 16  # bits
+FLASH_INTEGRITY_ECC_SIZE = 4  # bits
+FLASH_RELIABILITY_ECC_SIZE = 8  # bits
+PRINCE_NUM_HALF_ROUNDS = 5
 
-def _add_reliability_ecc(config: Dict[str, Any], in_val: int) -> str:
-    result, m = secded_gen.ecc_encode(config, "hamming", 68, in_val)
+# Create GF(2^64) with irreducible_polynomial = x^64 + x^4 + x^3 + x + 1
+GF_2_64 = ffield.FField(64,
+                        gen=((0x1 << 64) | (0x1 << 4) | (0x1 << 3) |
+                             (0x1 << 1) | 0x1))
 
-    m_nibbles = math.ceil((68 + m) / 4)
-    result = format(result, '0' + str(m_nibbles) + 'x')
 
-    # return full result
-    return result
+@functools.lru_cache(maxsize=None)
+def _xex_scramble(data: int, word_addr: int, flash_addr_key: int,
+                  flash_data_key: int) -> int:
+    operand_a = ((flash_addr_key & 0x0000ffffffffffff0000000000000000) >>
+                 48) | word_addr
```

Comment:
I think the code could be made more robust here if you would replace the `48` with something like `FLASH_WORD_SIZE - BANK_ADDR_WIDTH` and something similar for the first bit mask. I am just thinking in case we ever change the bank size or something.

Reply:
Fixed.